daml/Marketplace/Onboarding.daml

```diff
@@ -22,40 +61,50 @@ template UserSession
-          case role of
-            CustodianRole -> do
-              optCustodianInvitation <- lookupByKey @CustodianInvitation userKey
-              optCustodian <- lookupByKey @Custodian userKey
-              case (optCustodianInvitation, optCustodian) of
-                (None, None) -> void $ exerciseByKey @Operator operator Operator_OnboardCustodian with custodian = user
-                (Some custodianInvitationCid, Some custodianCid) -> archive custodianInvitationCid
-                _ -> return ()
-            IssuerRole -> do
-              optIssuerInvitation <- lookupByKey @IssuerInvitation userKey
-              optIssuer <- lookupByKey @Issuer userKey
-              case (optIssuerInvitation, optIssuer) of
-                (None, None) -> void $ exerciseByKey @Operator operator Operator_OnboardIssuer with issuer = user
-                (Some issuerInvitationCid, Some issuerCid) -> archive issuerInvitationCid
-                _ -> return ()
-            ExchangeRole -> do
-              optExchangeInvitation <- lookupByKey @ExchangeInvitation userKey
-              optExchange <- lookupByKey @Exchange userKey
-              case (optExchangeInvitation, optExchange) of
-                (None, None) -> void $ exerciseByKey @Operator operator Operator_OnboardExchange with exchange = user
-                (Some exchangeInvitationCid, Some exchangeCid) -> archive exchangeInvitationCid
-                _ -> return ()
-            InvestorRole -> do
-              optInvestorInvitation <- lookupByKey @InvestorInvitation userKey
-              optInvestor <- lookupByKey @Investor userKey
-              case (optInvestorInvitation, optInvestor) of
-                (None, None) -> void $ exerciseByKey @Operator operator Operator_OnboardInvestor with investor = user
-                (Some investorInvitationCid, Some investorCid) -> archive investorInvitationCid
-                _ -> return ()
-            BrokerRole -> do
-              optBrokerInvitation <- lookupByKey @BrokerInvitation userKey
-              optBroker <- lookupByKey @Broker userKey
-              case (optBrokerInvitation, optBroker) of
-                (None, None) -> void $ exerciseByKey @Operator operator Operator_OnboardBroker with broker = user
-                (Some brokerInvitationCid, Some brokerCid) -> archive brokerInvitationCid
-                _ -> return ()
+          create User with currentRole = role, ..
+          onboardRole role operator user
+
+template User
+  with
+    operator : Party
+    user : Party
+    currentRole : MarketRole
+  where
+    signatory operator
+
+    key (operator, user) : (Party, Party)
+    maintainer key._1
+
+    controller user can
+      nonconsuming User_RequestRoleChange : ContractId UserRoleChangeRequest
+        with
+          newRole : MarketRole
+        do
+          create UserRoleChangeRequest with ..
+
+    controller operator can
+      User_ChangeRole : ContractId User
```

Comment:
If you make the controller the `user`, the user would be able to call this choice directly instead of needing to create a `UserRoleChangeRequest`.

Unless it was a design decision to have it be a request, I think it would be preferable to reduce complexity and reduce the work the trigger needs to do.

Reply:
The request design was intentional because if the user wants to switch to a new role, the Operator needs to approve and then perform the `onboardRole` function to get them set up